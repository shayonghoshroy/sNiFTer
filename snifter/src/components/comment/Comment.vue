<template>
    <div>
        <va-card>
            <va-card-title>
                {{ sender }}
                <div v-if="this.user === this.sender && !this.isEditable && !isStatic" class="user-control">
                    <va-button-group>
                        <va-button @click="this.isEditable = true;" icon="edit"/>
                        <va-button @click="$emit('commentDeleted', { id })" icon="delete" />
                    </va-button-group>
                </div>
            </va-card-title>
            <va-card-content>
                <div class="comment-content">
                    <div v-if="!isEditable" class="rating">
                        <div class="rating-stars" v-for="n in parseInt(10)" :key="n">
                            <va-icon
                            :name="n <= latestRating ? 'star' : 'star_border'"
                            />
                        </div>
                    </div>
                    <div v-else class="rating">
                        <div class="rating-stars" v-for="n in parseInt(10)" :key="n">
                            <va-icon
                            v-if="!isEditable"
                            :name="n <= latestRating ? 'star' : 'star_border'"
                            />
                            <va-icon
                            v-else
                            :name="n <= latestRating ? 'star' : 'star_border'"
                            @click="latestRating = n"
                            />
                        </div>
                    </div>
                    <div class="message">
                        <p v-if="!isEditable" style="text-align: left;">{{ sanitizedMessage }}</p>
                        <div v-else>
                            <va-input
                            class="mb-4"
                            v-model="latestMessage"
                            :rules="[(v) => v.length <= 250 || `Comments must be 250 characters or less`]"
                            type="textarea"
                            style="width: 100%;"
                            />
                            <p style="text-align: left;">{{ latestMessage.length }} / 250</p>
                        </div>
                    </div>
                    <div class="comment-stub">
                        <div>
                            <va-icon
                            class="favorite-icon"
                            name="favorite_border"
                            />
                            {{ likes }}
                        </div>
                    </div>
                </div>
            </va-card-content>
            <va-card-actions v-if="isEditable" align="right">
                <va-button 
                style="text-align: right;"
                @click="latestMessage = message; isEditable = false;"
                >
                    Cancel
                </va-button>
                <va-button 
                style="text-align: right;"
                :disabled="disableCommentSubmit"
                @click="$emit('commentUpdated', { 'id': id, 'message': latestMessage, 'nft_rating': latestRating }); this.isEditable = false;"
                >
                    Update Review
                </va-button>
            </va-card-actions>
        </va-card>
    </div>
</template>

<script>
    export default {
        name: "Comment",
        props: {
            id: {
                type: Number,
                required: true, 
            },
            sender: {
                type: String,
                required: true,
            },
            message: {
                type: String,
                required: true,
            },
            collection: {
                type: String,
                required: true,
            },
            nft_id: {
                type: String,
                required: true,
            },
            nft_rating: {
                type: Number,
                required: true,
            },
            likes: {
                type: Number,
                default: 0,
            },
            user: {
                type: String,
                default: ''
            },
            isStatic: {
                type: Boolean,
                default: false
            }
        },
        mounted() {
            this.latestMessage = this.message;
            this.latestRating = this.nft_rating;
        },
        data() {
            return {
                isEditable: false,
                latestMessage: "",
                latestRating: 0
            }
        },
        computed: {
            sanitizedMessage() {
                var message = this.message;
                var sanitizedMessage = "";
                if(message.length > 50) {
                    while(message.length > 0) {
                        sanitizedMessage += message.substring(0, 50) + "\n";
                        message = message.substring(50);
                    }
                }
                else sanitizedMessage = message;

                return sanitizedMessage.trim();
            },
            disableCommentSubmit() {
                return this.latestMessage.length > 250;
            }
        },
    }
</script>

<style lang="css" scoped>

.user-control {
    position: absolute;
    right: 0;
    top: 0;
    padding: var(--va-card-padding);
}

.comment-content {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: center;
    padding: 10px;
}

.rating {
    display: flex;
    justify-content: center;
}

.favorite-icon:hover {
    cursor: pointer;
}

.message {
    width: 100%;
}

</style>
