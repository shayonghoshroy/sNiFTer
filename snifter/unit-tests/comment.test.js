import { mount } from '@vue/test-utils'
import Comment from '../src/components/comment/Comment.vue'

// test component by mounting it with a fake data object

describe('Comment.test.js', () => {
  let cmp

  beforeEach(() => {
    cmp = mount(Comment, {
      propsData: {
        id: 1,
        sender: "shyguy",
        message: "I am a comment.",
        collection: '0x79fcdef22feed20eddacbb2587640e45491b757f',
        nft_id: '0x79fcdef22feed20eddacbb2587640e45491b757f-9973',
        nft_rating: 10,
        likes: 0,
        user: "shyguy",
        isStatic: false,
      }
    })
  })

  // check if the component has expected data

  it('got comment id', () => {
    expect(cmp.vm.id).toEqual(1)
  })

  it('got comment sender', () => {
    expect(cmp.vm.sender).toEqual("shyguy")
  })

  it('got comment message', () => {
    expect(cmp.vm.message).toEqual("I am a comment.")
  })

  it('got comment collection', () => {
    expect(cmp.vm.collection).toEqual('0x79fcdef22feed20eddacbb2587640e45491b757f')
  })

  it('got comment nft_id', () => {
    expect(cmp.vm.nft_id).toEqual('0x79fcdef22feed20eddacbb2587640e45491b757f-9973')
  })

  it('got comment nft_rating', () => {
    expect(cmp.vm.nft_rating).toEqual(10)
  })

  it('got comment likes', () => {
    expect(cmp.vm.likes).toEqual(0)
  })

  it('got comment user', () => {
    expect(cmp.vm.user).toEqual("shyguy")
  })

  it('got comment isStatic', () => {
    expect(cmp.vm.isStatic).toEqual(false)
  })

  it('has the expected html structure', () => {
    expect(cmp.element).toMatchSnapshot()
  })
})
