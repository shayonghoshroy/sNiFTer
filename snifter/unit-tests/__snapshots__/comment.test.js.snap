// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment.test.js has the expected html structure 1`] = `
<div>
  <va-card>
    <va-card-title>
      shyguy 
      <div
        class="user-control"
      >
        <va-button-group>
          <va-button
            icon="edit"
          />
          <va-button
            icon="delete"
          />
        </va-button-group>
      </div>
    </va-card-title>
    <va-card-content>
      <div
        class="comment-content"
      >
        <div
          class="rating"
        >
          
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          <div
            class="rating-stars"
          >
            <va-icon
              name="star"
            />
          </div>
          
        </div>
        <div
          class="message"
        >
          <p
            style="text-align: left;"
          >
            I am a comment.
          </p>
        </div>
      </div>
    </va-card-content>
    <!--v-if-->
  </va-card>
</div>
`;
